concordium-contracts-common/src/schema_json.rs

```diff
@@ -953,192 +1472,347 @@ fn deserial_string<R: Read>(source: &mut R, size_len: SizeLength) -> ParseResult
+                });
+
+                owned_receive_name
+                    .map(|orn| {
+                        let receive_name = orn.as_receive_name();
+                        let contract_name = receive_name.contract_name();
+                        let func_name = receive_name.entrypoint_name();
+                        json!({"contract": contract_name, "func": func_name})
+                    })
+                    .map_err(|e| deserial_error(e.0))
             }
             Type::ULeb128(constraint) => {
-                let int = deserial_biguint(source, *constraint)?;
+                let int = deserial_biguint(source, *constraint).map_err(|_| {
+                    deserial_error("Could not parse unsigned integer from value".into())
+                })?;
                 Ok(Value::String(int.to_string()))
             }
             Type::ILeb128(constraint) => {
-                let int = deserial_bigint(source, *constraint)?;
+                let int = deserial_bigint(source, *constraint).map_err(|_| {
+                    deserial_error("Could not parse signed integer from value".into())
+                })?;
                 Ok(Value::String(int.to_string()))
             }
             Type::ByteList(size_len) => {
-                let len = deserial_length(source, *size_len)?;
-                let mut string =
-                    String::with_capacity(std::cmp::min(MAX_PREALLOCATED_CAPACITY, 2 * len));
-                for _ in 0..len {
-                    let byte = source.read_u8()?;
-                    string.push_str(&format!("{:02x?}", byte));
+                let len = deserial_length(source, *size_len).map_err(|_| {
+                    ParseErrorWithReason("Could not parse ByteList length from value".into())
+                });
+                let bytes: core::result::Result<Vec<_>, _> =
+                    len.as_ref().map_err(|e| e.to_owned()).and_then(|len| {
+                        let mut bytes =
+                            Vec::with_capacity(std::cmp::min(MAX_PREALLOCATED_CAPACITY, *len));
+
+                        for i in 0..*len {
+                            let byte = source.read_u8().map_err(|_| {
+                                ParseErrorWithReason(format!(
+                                    "Failed to read byte {} of ByteList value",
+                                    i
+                                ))
+                            });
+                            bytes.push(byte);
+                        }
+
+                        bytes.into_iter().collect()
+                    });
+
+                match (len, bytes) {
+                    (Ok(len), Ok(bytes)) => {
+                        let mut string = String::with_capacity(std::cmp::min(
+                            MAX_PREALLOCATED_CAPACITY,
+                            2 * len,
+                        ));
+                        bytes.into_iter().for_each(|b| string.push_str(&format!("{:02x?}", b)));
```

Comment:
I am really not a fan of these effectful closures. I would strongly prefer to keep the for loop that we had.